File: util/benchmark.py
#!./util/env/bin/python3

from os.path import join
from subprocess import Popen, PIPE
import sys


def run_trial(interpreter, benchmark):
  """Runs the benchmark once and returns the elapsed time."""
  args = [interpreter, join('test', 'benchmark', benchmark + '.lox')]
  proc = Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE)

  out, err = proc.communicate()

  out = out.decode("utf-8").replace('\r\n', '\n')

  # Remove the trailing last empty line.
  out_lines = out.split('\n')
  if out_lines[-1] == '':
    del out_lines[-1]

  # The benchmark should print the elapsed time last.
  return float(out_lines[-1])


def run_comparison(interpreters, benchmark):
  trial = 1
  best = {}
  for interpreter in interpreters:
    best[interpreter] = 9999

  while True:
    for interpreter in interpreters:
      elapsed = run_trial(interpreter, benchmark)
      if elapsed < best[interpreter]:
        best[interpreter] = elapsed

    best_time = 999
    worst_time = 0
    for interpreter in interpreters:
      if best[interpreter] < best_time:
        best_time = best[interpreter]
        best_interpreter = interpreter
      if best[interpreter] > worst_time:
        worst_time = best[interpreter]

    # Turn the time measurement into an effort measurement in units where 1
    # "work" is just the total thing the benchmark does.
    worst_work = 1.0 / worst_time

    print("trial #{0}".format(trial))
    for interpreter in interpreters:
      if interpreter == best_interpreter:
        best_work = 1.0 / best[interpreter]
        work_ratio = best_work / worst_work
        faster = 100 * (work_ratio - 1.0)
        suffix = "{0:0.4f}% faster".format(faster)
      else:
        ratio = best[interpreter] / best_time
        suffix = "{0:0.4}x time of best".format(ratio)
      print("  {0}   best {1:0.4}s  {2}".format(
          interpreter, best[interpreter], suffix))

    trial += 1


def run_benchmark(interpreter, benchmark):
  trial = 1
  best = 9999

  while True:
    elapsed = run_trial(interpreter, benchmark)
    if elapsed < best:
      best = elapsed

    print("trial #{0}  {1}   best {2:0.4}s".format(trial, interpreter, best))
    trial += 1


interpreters = ['build/clox']
if len(sys.argv) == 2:
  benchmark = sys.argv[1]
elif len(sys.argv) > 2:
  interpreters = sys.argv[1:-1]
  benchmark = sys.argv[-1]
else:
  print('Usage: benchmark.py [interpreters...] <benchmark>')
  sys.exit(1)

if len(interpreters) > 1:
  run_comparison(interpreters, benchmark)
else:
  run_benchmark(interpreters[0], benchmark)

File: tool/bin/benchmark.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

void main(List<String> arguments) {
  if (arguments.isEmpty) {
    print('Usage: benchmark.py [interpreters...] <benchmark>');
    exit(1);
  }

  var interpreters = ['build/clox'];
  var benchmark = arguments.last;
  if (arguments.length > 1) {
    interpreters = arguments.sublist(0, arguments.length - 1);
  }

  if (interpreters.length > 1) {
    runComparison(interpreters, benchmark);
  } else {
    runBenchmark(interpreters[0], benchmark);
  }
}

void runBenchmark(String interpreter, String benchmark) {
  var trial = 1;
  var best = 9999.0;

  for (;;) {
    var elapsed = runTrial(interpreter, benchmark);
    if (elapsed < best) best = elapsed;

    var bestSeconds = best.toStringAsFixed(2);
    print("trial #$trial  $interpreter   best ${bestSeconds}s");
    trial++;
  }
}

/// Runs the benchmark once and returns the elapsed time.
double runTrial(String interpreter, String benchmark) {
  var result = Process.runSync(
      interpreter, [p.join("test", "benchmark", "$benchmark.lox")]);
  var outLines = const LineSplitter().convert(result.stdout as String);

  // Remove the trailing last empty line.
  if (outLines.last == "") outLines.removeLast();

  // The benchmark should print the elapsed time last.
  return double.parse(outLines.last);
}

void runComparison(List<String> interpreters, String benchmark) {
  var trial = 1;
  var best = {for (var interpreter in interpreters) interpreter: 9999.0};

  for (;;) {
    for (var interpreter in interpreters) {
      var elapsed = runTrial(interpreter, benchmark);
      if (elapsed < best[interpreter]) best[interpreter] = elapsed;
    }

    var bestTime = 999.0;
    var worstTime = 0.0;
    String bestInterpreter;
    for (var interpreter in interpreters) {
      if (best[interpreter] < bestTime) {
        bestTime = best[interpreter];
        bestInterpreter = interpreter;
      }
      if (best[interpreter] > worstTime) {
        worstTime = best[interpreter];
      }
    }

    // Turn the time measurement into an effort measurement in units where 1
    // "work" is just the total thing the benchmark does.
    var worstWork = 1.0 / worstTime;

    print("trial #$trial");
    for (var interpreter in interpreters) {
      String suffix;
      if (interpreter == bestInterpreter) {
        var bestWork = 1.0 / best[interpreter];
        var workRatio = bestWork / worstWork;
        var faster = 100 * (workRatio - 1.0);
        suffix = "${faster.toStringAsFixed(4)}% faster";
      } else {
        var ratio = best[interpreter] / bestTime;
        suffix = "${ratio.toStringAsFixed(4)}x time of best";
      }
      var bestString = best[interpreter].toStringAsFixed(4);
      print("  ${interpreter.padRight(30)}   best ${bestString}s  $suffix");
    }

    trial++;
  }
}
